src/js/forms.js

```diff
@@ -1,65 +1,53 @@
+import Input from './input';
+import State from './state';
+
 class Forms {
 	/**
 	* Class constructor.
 	* @param {HTMLElement} [formElement] - The form element in the DOM
 	* @param {Object} [options={}] - An options object for configuring the form
 	*/
 	constructor(formElement, options) {
-		this.formEl = formElement;
-		this.inputFields = Array.from(formElement.querySelectorAll('input', 'select', 'textarea'));
+		if (formElement.nodeName !== 'FORM') {
+			throw new Error(`[data-o-component="o-forms"] must be set on a form element. It is currently set on a '${formElement.nodeName.toLowerCase()}'.`);
+		}
+
+		this.form = formElement;
+		this.formElements = Array.from(this.form.elements, element => new Input(element));
+		this.stateArray = [];
 
 		this.opts = Object.assign({
-			useBrowserValidation: true
+			useBrowserValidation: false
 		}, options);
 
-		this.className = {
-			invalid: 'o-forms-input--invalid',
-			valid: 'o-forms-input--valid'
-		};
-
 		if (!this.opts.useBrowserValidation) {
-			this.formEl.setAttribute('novalidate', true);
-			this.formEl.addEventListener('submit', this);
+			this.form.setAttribute('novalidate', true);
+			this.form.addEventListener('submit', this);
 		} else {
-			this.formEl.removeAttribute('novalidate');
-			let submit = this.formEl.querySelector('input[type=submit]');
-			submit.addEventListener('click', this);
-			submit.addEventListener('keydown', this);
+			this.form.removeAttribute('novalidate');
+			let submits = this.form.querySelectorAll('[type=submit]');
+			submits.forEach(submit => {
+				submit.addEventListener('click', this);
+				submit.addEventListener('keydown', this);
+			});
 		}
-
-		this.inputFields.forEach(input => {
-			input.addEventListener('blur', this);
-			input.addEventListener('input', this);
-		});
 	}
 
 	/**
 	 * Event Handler
 	 * @param {Object} event - The event emitted by element/window interactions
 	 */
 	handleEvent(e) {
-		let field = e.target.closest('.o-forms-input');
-		if (e.type === 'blur') {
-			field.querySelectorAll('input').forEach(input => this.validate(input));
-		}
-
-		if (e.type === 'input') {
-			if (e.target.validity.valid && field.classList.contains(this.className.invalid)) {
-				field.classList.replace(this.className.invalid, this.className.valid);
-			}
-		}
-
 		const RETURN_KEY = 13;
 		if (e.type === 'click' || (e.type === 'keydown' && e.key === RETURN_KEY)) {
-			if (!this.formEl.reportValidity()) {
-				this.inputFields.forEach(input => this.validate(input));
+			if (!this.form.checkValidity()) {
```

Comment:
Does `checkValidity` exist in IE? MDN doesn't say yes or no 😔 
![image](https://user-images.githubusercontent.com/1569131/56646905-66e67a00-6678-11e9-8162-bb72cd851213.png)

Reply:
it seems to, as far as IE10 https://caniuse.com/#search=reportV. Perhaps forms should default to using browser validation for <IE10